casper/src/main/scala/coop/rchain/casper/LastFinalizedHeightConstraintChecker.scala

```diff
@@ -1,52 +1,54 @@
 package coop.rchain.casper
 
 import cats.effect.Sync
-import cats.syntax.applicative._
-import cats.syntax.flatMap._
-import cats.syntax.functor._
+import cats.syntax.all._
+import com.google.protobuf.ByteString
 import coop.rchain.blockstorage.BlockStore
-import coop.rchain.blockstorage.dag.BlockDagRepresentation
 import coop.rchain.blockstorage.finality.LastFinalizedStorage
-import coop.rchain.blockstorage.syntax._
+import coop.rchain.casper.blocks.proposer.{
+  CheckProposeConstraintsResult,
+  TooFarAheadOfLastFinalized
+}
 import coop.rchain.casper.protocol.BlockMessage
-import coop.rchain.models.Validator.Validator
+import coop.rchain.casper.syntax._
 import coop.rchain.shared.Log
 
-final class LastFinalizedHeightConstraintChecker[F[_]: Sync: LastFinalizedStorage: Log](
-    heightConstraintThreshold: Long
-) {
+final class LastFinalizedHeightConstraintChecker[F[_]: Sync: LastFinalizedStorage: Log] {
   def check(
-      dag: BlockDagRepresentation[F],
+      s: CasperSnapshot[F],
+      // TODO having genesis is a weird way to check, remove
       genesis: BlockMessage,
-      validator: Validator
-  ): F[Boolean] =
+      validatorIdentity: ValidatorIdentity
+  ): F[CheckProposeConstraintsResult] = {
+    val validator                 = ByteString.copyFrom(validatorIdentity.publicKey.bytes)
+    val heightConstraintThreshold = s.onChainState.shardConf.heightConstraintThreshold
     for {
       lastFinalizedBlockHash <- LastFinalizedStorage[F].getOrElse(genesis.blockHash)
-      lastFinalizedBlock     <- dag.lookupUnsafe(lastFinalizedBlockHash)
-      latestMessageOpt       <- dag.latestMessage(validator)
+      lastFinalizedBlock     <- s.dag.lookupUnsafe(lastFinalizedBlockHash)
+      latestMessageOpt       <- s.dag.latestMessage(validator)
       result <- latestMessageOpt match {
                  case Some(latestMessage) =>
                    val latestFinalizedHeight = lastFinalizedBlock.blockNum
                    val heightDifference      = latestMessage.blockNum - latestFinalizedHeight
+                   val result =
+                     if (heightDifference <= heightConstraintThreshold)
+                       CheckProposeConstraintsResult.success
+                     else TooFarAheadOfLastFinalized
                    Log[F].info(
                      s"Latest message is $heightDifference blocks ahead of the last finalized block"
-                   ) >> (heightDifference <= heightConstraintThreshold).pure[F]
+                   ) >> result.pure[F]
                  case None =>
-                   Sync[F].raiseError[Boolean](
-                     new IllegalStateException("Validator does not have a latest message")
-                   )
+                   CheckProposeConstraintsResult.success.pure[F]
+//                   Sync[F].raiseError[CheckProposeConstraintsResult](
+//                     new IllegalStateException("Validator does not have a latest message")
+//                   )
```

Comment:
I think this error message is misleading and does not tell much, as it is not related to this particular constraint checker.   If validator is bonded, but had not issued its first message - it should be able to create one, and peers should be able to validate the very first message. **[1]**

If it is not bonded - propose status should return exactly this **[2]**. And here is where all propose constraints check happen. Active validator check should fail.
https://github.com/rchain/rchain/blob/bc95a9cc30333f465939f660c4f16de8415a1492/casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala#L82-L97 

**[1]** *We had a related discussion - when to create latest message, should that be message where bonding happened or the very first block proposed. I believe it should be in the second one, but in both cases, having this error raised does not make sense imo. If no latest message seen for validator - network should not restrict it to create its first message.*

**[2]** Only propose is mentioned here because checking of propose constraint during validation is not implemented - its some future task,  which is part of shard config feature. Boilerplate code is mostly ready, but <https://github.com/rchain/rchain/issues/3350> feature is needed as well. But when validating, the reasoning remains the same imo.

Reply:
@nzpr from our discussion on standup, for now it makes sense to keep throwing error when validator has no latest message. Please add this change as part of PR #3355.

Reply:
https://github.com/rchain/rchain/pull/3355#issuecomment-813358683